truffle/src/main/java/org/jruby/truffle/core/string/StringNodes.java

```diff
@@ -1355,6 +1360,153 @@ private int prevCharHead(Encoding enc, byte[]bytes, int p, int s, int end) {
+
+            final byte[] pBytes = rope.getBytes();
+            final int e = pBytes.length;
+
+            int p = 0;
+            int p1 = 0;
+
+            p = StringSupport.searchNonAscii(pBytes, p, e);
+            if (p == -1) {
+                p = e;
+            }
+            while (p < e) {
+                int ret = enc.length(pBytes, p, e);
+                if (MBCLEN_NEEDMORE_P(ret)) {
+                    break;
+                } else if (MBCLEN_CHARFOUND_P(ret)) {
+                    p += MBCLEN_CHARFOUND_LEN(ret);
+                } else if (MBCLEN_INVALID_P(ret)) {
+                    // p1~p: valid ascii/multibyte chars
+                    // p ~e: invalid bytes + unknown bytes
+                    int clen = enc.maxLength();
+                    if (p1 < p) {
+                        buf = makeConcatNode.executeMake(buf, RopeOperations.create(ArrayUtils.extractRange(pBytes, p1, p), enc, CodeRange.CR_VALID), enc);
+                    }
+
+                    if (e - p < clen) {
+                        clen = e - p;
+                    }
+                    if (clen <= 2) {
+                        clen = 1;
+                    } else {
+                        final int q = p;
+                        clen--;
+                        for (; clen > 1; clen--) {
+                            ret = enc.length(pBytes, q, q + clen);
+                            if (MBCLEN_NEEDMORE_P(ret)) {
+                                break;
+                            } else if (MBCLEN_INVALID_P(ret)) {
+                                continue;
+                            }
+                        }
+                    }
+                    Rope invalid = RopeOperations.create(ArrayUtils.extractRange(pBytes, p, p + clen), enc, CodeRange.CR_BROKEN);
+                    DynamicObject repl = (DynamicObject) yield(frame, block, createString(invalid));
+                    buf = makeConcatNode.executeMake(buf, rope(repl), enc);
+                    p += clen;
+                    p1 = p;
+                    p = StringSupport.searchNonAscii(pBytes, p, e);
+                    if (p == -1) {
+                        p = e;
+                        break;
+                    }
+                }
+            }
+            if (p1 < p) {
+                buf = makeConcatNode.executeMake(buf, RopeOperations.create(ArrayUtils.extractRange(pBytes, p1, p), enc, CodeRange.CR_VALID), enc);
+            }
+            if (p < e) {
+                Rope invalid = RopeOperations.create(ArrayUtils.extractRange(pBytes, p, e), enc, CodeRange.CR_BROKEN);
+                DynamicObject repl = (DynamicObject) yield(frame, block, createString(invalid));
+                buf = makeConcatNode.executeMake(buf, rope(repl), enc);
+            }
+
+            return createString(buf);
+        }
+
+        @Specialization(guards = { "!is7Bit(string)", "!isValidCodeRange(string)", "!isAsciiCompatible(string)" })
```

Comment:
Replace `"!is7Bit(string)", "!isValidCodeRange(string)"` with `isBrokenCodeRange(string)`. We never have ropes with unknown code ranges floating around, so we don't need to worry about that case. When you eliminate the other two, broken code ranges are all that remain. Incidentally, I think it also clarifies the intent more.